```
2020-05-25T08:09:22.3713888Z Caused by: org.apache.maven.plugin.MojoFailureException: File '/home/runner/work/quarkus/quarkus/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InterceptorInfo.java' has not been previously formatted.  Please format file and commit before running validation!
```

:wink:

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BuildTimeEnabledProcessor.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BuildTimeEnabledProcessor.java
@@ -10,6 +10,7 @@ import org.eclipse.microprofile.config.Config;
 import org.eclipse.microprofile.config.spi.ConfigProviderResolver;
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationTarget;
+import org.jboss.jandex.AnnotationTarget.Kind;
 import org.jboss.jandex.AnnotationValue;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.FieldInfo;
@@ -138,11 +139,6 @@ public class BuildTimeEnabledProcessor {
 
         // the transformer just tries to match targets and then enables or disables the bean accordingly
         annotationsTransformer.produce(new AnnotationsTransformerBuildItem(new AnnotationsTransformer() {
-            @Override
-            public boolean appliesTo(AnnotationTarget.Kind kind) {
-                return kind == AnnotationTarget.Kind.METHOD || kind == AnnotationTarget.Kind.CLASS
-                        || kind == AnnotationTarget.Kind.FIELD;
-            }
 
             @Override
             public void transform(TransformationContext ctx) {
@@ -150,18 +146,18 @@ public class BuildTimeEnabledProcessor {
                 if (ctx.isClass()) {
                     DotName classDotName = target.asClass().name();
                     if (classTargets.containsKey(classDotName)) {
-                        transformBean(ctx.transform(), classTargets.get(classDotName));
+                        transformBean(target, ctx.transform(), classTargets.get(classDotName));
                     }
                 } else if (ctx.isMethod()) {
                     MethodInfo method = target.asMethod();
                     if (methodTargets.containsKey(method)) {
-                        transformBean(ctx.transform(), methodTargets.get(method));
+                        transformBean(target, ctx.transform(), methodTargets.get(method));
                     }
                 } else if (ctx.isField()) {
                     FieldInfo field = target.asField();
                     String uniqueFieldName = toUniqueString(field);
                     if (fieldTargets.containsKey(uniqueFieldName)) {
-                        transformBean(ctx.transform(), fieldTargets.get(uniqueFieldName));
+                        transformBean(target, ctx.transform(), fieldTargets.get(uniqueFieldName));
                     }
                 }
             }
@@ -172,11 +168,11 @@ public class BuildTimeEnabledProcessor {
         return field.declaringClass().name().toString() + "." + field.name();
     }
 
-    private void transformBean(Transformation transform, boolean enable) {
+    private void transformBean(AnnotationTarget target, Transformation transform, boolean enable) {
         if (enable) {
             enableBean(transform);
         } else {
-            disableBean(transform);
+            disableBean(target, transform);
         }
         transform.done();
     }
@@ -191,7 +187,13 @@ public class BuildTimeEnabledProcessor {
         return AnnotationValue.createIntegerValue("value", Integer.MAX_VALUE - 1);
     }
 
-    private void disableBean(Transformation transform) {
-        transform.add(DotNames.ALTERNATIVE);
+    private void disableBean(AnnotationTarget target, Transformation transform) {
+        if (target.kind() == Kind.CLASS) {
+            // Veto the class
+            transform.add(DotNames.VETOED);
+        } else {
+            // Add @Alternative to the producer
+            transform.add(DotNames.ALTERNATIVE);
+        }
     }
 }

--- a/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/profile/IfBuildProfileTest.java
+++ b/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/profile/IfBuildProfileTest.java
@@ -1,14 +1,21 @@
 package io.quarkus.arc.test.profile;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.Priority;
 import javax.enterprise.context.ApplicationScoped;
 import javax.enterprise.inject.Instance;
 import javax.enterprise.inject.Produces;
 import javax.enterprise.inject.spi.CDI;
 import javax.inject.Inject;
 import javax.inject.Singleton;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
 
 import org.jboss.shrinkwrap.api.ShrinkWrap;
 import org.jboss.shrinkwrap.api.spec.JavaArchive;
@@ -25,7 +32,8 @@ public class IfBuildProfileTest {
     static final QuarkusUnitTest config = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
                     .addClasses(Producer.class, OtherProducer.class, AnotherProducer.class,
-                            GreetingBean.class, Hello.class, PingBean.class, PongBean.class, FooBean.class, BarBean.class));
+                            GreetingBean.class, Hello.class, PingBean.class, PongBean.class, FooBean.class, BarBean.class,
+                            TestInterceptor.class, ProdInterceptor.class, Logging.class));
 
     @Inject
     Hello hello;
@@ -35,11 +43,15 @@ public class IfBuildProfileTest {
 
     @Test
     public void testInjection() {
+        assertFalse(TestInterceptor.INTERCEPTED.get());
+        assertFalse(ProdInterceptor.INTERCEPTED.get());
         assertEquals("hello from test. Foo is: foo from test", hello.hello());
         assertEquals("ping", hello.ping());
         assertEquals("pong", hello.pong());
         assertEquals("foo from test", hello.foo());
         assertTrue(barBean.isUnsatisfied());
+        assertTrue(TestInterceptor.INTERCEPTED.get());
+        assertFalse(ProdInterceptor.INTERCEPTED.get());
     }
 
     @Test
@@ -47,6 +59,7 @@ public class IfBuildProfileTest {
         assertEquals("hello from test. Foo is: foo from test", CDI.current().select(GreetingBean.class).get().greet());
     }
 
+    @Logging
     @ApplicationScoped
     static class Hello {
 
@@ -198,5 +211,37 @@ public class IfBuildProfileTest {
                 }
             };
         }
+    }
+
+    @IfBuildProfile("test")
+    @Priority(1)
+    @Interceptor
+    @Logging
+    static class TestInterceptor {
+
+        static final AtomicBoolean INTERCEPTED = new AtomicBoolean(false);
+
+        @AroundInvoke
+        public Object aroundInvoke(InvocationContext ctx) throws Exception {
+            INTERCEPTED.set(true);
+            return ctx.proceed();
+        }
+
+    }
+
+    @IfBuildProfile("prod")
+    @Priority(10)
+    @Interceptor
+    @Logging
+    static class ProdInterceptor {
+
+        static final AtomicBoolean INTERCEPTED = new AtomicBoolean(false);
+
+        @AroundInvoke
+        public Object aroundInvoke(InvocationContext ctx) throws Exception {
+            INTERCEPTED.set(true);
+            return ctx.proceed();
+        }
+
     }
 }

--- a/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/profile/Logging.java
+++ b/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/profile/Logging.java
@@ -1,0 +1,19 @@
+package io.quarkus.arc.test.profile;
+
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import javax.interceptor.InterceptorBinding;
+
+@Target({ TYPE, METHOD })
+@Retention(RUNTIME)
+@Documented
+@InterceptorBinding
+public @interface Logging {
+
+}

--- a/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/profile/UnlessBuildProfileTest.java
+++ b/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/profile/UnlessBuildProfileTest.java
@@ -1,14 +1,21 @@
 package io.quarkus.arc.test.profile;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.Priority;
 import javax.enterprise.context.ApplicationScoped;
 import javax.enterprise.inject.Instance;
 import javax.enterprise.inject.Produces;
 import javax.enterprise.inject.spi.CDI;
 import javax.inject.Inject;
 import javax.inject.Singleton;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
 
 import org.jboss.shrinkwrap.api.ShrinkWrap;
 import org.jboss.shrinkwrap.api.spec.JavaArchive;
@@ -25,7 +32,8 @@ public class UnlessBuildProfileTest {
     static final QuarkusUnitTest config = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
                     .addClasses(Producer.class, AnotherProducer.class,
-                            GreetingBean.class, Hello.class, PingBean.class, PongBean.class, FooBean.class, BarBean.class));
+                            GreetingBean.class, Hello.class, PingBean.class, PongBean.class, FooBean.class, BarBean.class,
+                            TestInterceptor.class, ProdInterceptor.class, Logging.class));
 
     @Inject
     Hello hello;
@@ -35,11 +43,15 @@ public class UnlessBuildProfileTest {
 
     @Test
     public void testInjection() {
+        assertFalse(TestInterceptor.INTERCEPTED.get());
+        assertFalse(ProdInterceptor.INTERCEPTED.get());
         assertEquals("hello from not prod. Foo is: foo from not prod", hello.hello());
         assertEquals("ping", hello.ping());
         assertEquals("pong from not prod", hello.pong());
         assertEquals("foo from not prod", hello.foo());
         assertTrue(barBean.isUnsatisfied());
+        assertTrue(TestInterceptor.INTERCEPTED.get());
+        assertFalse(ProdInterceptor.INTERCEPTED.get());
     }
 
     @Test
@@ -47,6 +59,7 @@ public class UnlessBuildProfileTest {
         assertEquals("hello from not prod. Foo is: foo from not prod", CDI.current().select(GreetingBean.class).get().greet());
     }
 
+    @Logging
     @ApplicationScoped
     static class Hello {
 
@@ -171,5 +184,37 @@ public class UnlessBuildProfileTest {
                 }
             };
         }
+    }
+
+    @UnlessBuildProfile("prod")
+    @Priority(1)
+    @Interceptor
+    @Logging
+    static class TestInterceptor {
+
+        static final AtomicBoolean INTERCEPTED = new AtomicBoolean(false);
+
+        @AroundInvoke
+        public Object aroundInvoke(InvocationContext ctx) throws Exception {
+            INTERCEPTED.set(true);
+            return ctx.proceed();
+        }
+
+    }
+
+    @UnlessBuildProfile("test")
+    @Priority(10)
+    @Interceptor
+    @Logging
+    static class ProdInterceptor {
+
+        static final AtomicBoolean INTERCEPTED = new AtomicBoolean(false);
+
+        @AroundInvoke
+        public Object aroundInvoke(InvocationContext ctx) throws Exception {
+            INTERCEPTED.set(true);
+            return ctx.proceed();
+        }
+
     }
 }

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
@@ -996,6 +996,10 @@ public class BeanDeployment {
         }
         List<InterceptorInfo> interceptors = new ArrayList<>();
         for (ClassInfo interceptorClass : interceptorClasses) {
+            if (annotationStore.hasAnnotation(interceptorClass, DotNames.VETOED)) {
+                // Skip vetoed interceptors
+                continue;
+            }
             interceptors
                     .add(Interceptors.createInterceptor(interceptorClass, this, injectionPointTransformer, annotationStore));
         }

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InterceptorInfo.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InterceptorInfo.java
@@ -11,12 +11,15 @@ import org.jboss.jandex.DotName;
 import org.jboss.jandex.MethodInfo;
 import org.jboss.jandex.Type;
 import org.jboss.jandex.Type.Kind;
+import org.jboss.logging.Logger;
 
 /**
  *
  * @author Martin Kouba
  */
-class InterceptorInfo extends BeanInfo implements Comparable<InterceptorInfo> {
+public class InterceptorInfo extends BeanInfo implements Comparable<InterceptorInfo> {
+
+    private static final Logger LOGGER = Logger.getLogger(InterceptorInfo.class);
 
     private final Set<AnnotationInstance> bindings;
 
@@ -77,13 +80,16 @@ class InterceptorInfo extends BeanInfo implements Comparable<InterceptorInfo> {
         this.aroundConstruct = aroundConstruct;
         this.postConstruct = postConstruct;
         this.preDestroy = preDestroy;
+        if (aroundConstruct == null && aroundInvoke == null && preDestroy == null && postConstruct == null) {
+            LOGGER.warnf("%s declares no around-invoke method nor a lifecycle callback!", this);
+        }
     }
 
-    Set<AnnotationInstance> getBindings() {
+    public Set<AnnotationInstance> getBindings() {
         return bindings;
     }
 
-    int getPriority() {
+    public int getPriority() {
         return priority;
     }
 
@@ -103,7 +109,7 @@ class InterceptorInfo extends BeanInfo implements Comparable<InterceptorInfo> {
         return preDestroy;
     }
 
-    boolean intercepts(InterceptionType interceptionType) {
+    public boolean intercepts(InterceptionType interceptionType) {
         switch (interceptionType) {
             case AROUND_INVOKE:
                 return aroundInvoke != null;

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Interceptors.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Interceptors.java
@@ -1,5 +1,6 @@
 package io.quarkus.arc.processor;
 
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Set;
 import javax.enterprise.inject.spi.DefinitionException;
@@ -46,7 +47,9 @@ final class Interceptors {
             LOGGER.info("An interceptor " + interceptorClass + " does not declare any @Priority. " +
                     "It will be assigned a default priority value of 0.");
         }
-        return new InterceptorInfo(interceptorClass, beanDeployment, bindings,
+        return new InterceptorInfo(interceptorClass, beanDeployment,
+                bindings.size() == 1 ? Collections.singleton(bindings.iterator().next())
+                        : Collections.unmodifiableSet(bindings),
                 Injection.forBean(interceptorClass, null, beanDeployment, transformer), priority);
     }
 

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/vetoed/Logging.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/vetoed/Logging.java
@@ -1,0 +1,18 @@
+package io.quarkus.arc.test.vetoed;
+
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import javax.interceptor.InterceptorBinding;
+
+@Target({ TYPE, METHOD })
+@Retention(RUNTIME)
+@Documented
+@InterceptorBinding
+public @interface Logging {
+
+}

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/vetoed/VetoedTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/vetoed/VetoedTest.java
@@ -8,15 +8,20 @@ import io.quarkus.arc.Arc;
 import io.quarkus.arc.ArcContainer;
 import io.quarkus.arc.test.ArcTestContainer;
 import java.util.AbstractList;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Priority;
 import javax.enterprise.context.Dependent;
 import javax.enterprise.inject.Vetoed;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
 public class VetoedTest {
 
     @RegisterExtension
-    public ArcTestContainer container = new ArcTestContainer(Seven.class, One.class);
+    public ArcTestContainer container = new ArcTestContainer(Seven.class, One.class, VetoedInterceptor.class, Logging.class);
 
     @Test
     public void testVetoed() {
@@ -24,9 +29,13 @@ public class VetoedTest {
         assertTrue(arc.instance(Seven.class).isAvailable());
         // One is vetoed
         assertFalse(arc.instance(One.class).isAvailable());
+        assertFalse(VetoedInterceptor.INTERCEPTED.get());
         assertEquals(Integer.valueOf(7), Integer.valueOf(arc.instance(Seven.class).get().size()));
+        // Interceptor is vetoed
+        assertFalse(VetoedInterceptor.INTERCEPTED.get());
     }
 
+    @Logging
     @Dependent
     static class Seven extends AbstractList<Integer> {
 
@@ -54,6 +63,22 @@ public class VetoedTest {
         @Override
         public int size() {
             return 1;
+        }
+
+    }
+
+    @Vetoed
+    @Priority(1)
+    @Interceptor
+    @Logging
+    static class VetoedInterceptor {
+
+        static final AtomicBoolean INTERCEPTED = new AtomicBoolean(false);
+
+        @AroundInvoke
+        public Object aroundInvoke(InvocationContext ctx) throws Exception {
+            INTERCEPTED.set(true);
+            return ctx.proceed();
         }
 
     }

